urldeb33fd2fa_friedemann_distripedia_tgz-pJ8-org_json_TestJ8/src/websocket/client/DistripediaClient.java

```diff
@@ -0,0 +1,368 @@
+import java.util.Collection;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import websocket.client.WebSocket.CloseEvent;
+import websocket.client.WebSocket.CloseHandler;
+import websocket.client.WebSocket.MessageEvent;
+import websocket.client.WebSocket.MessageHandler;
+import websocket.client.WebSocket.OpenEvent;
+import websocket.client.WebSocket.OpenHandler;
+import websocket.common.Logger;
+import websocket.common.Message;
+
+/**
+ * A Java Implementation of the Distripedia JS-Client.
+ * 
+ * @author David
+ */
+public class DistripediaClient implements OpenHandler, CloseHandler, MessageHandler
+{
+	private final WebSocket webSocket;
+	private final LocalStorage localStorage;
+	
+	private ArticleReceivedHandler articleReceivedHandler;
+
+	private boolean debug = false;
+	private boolean isCachedArticlesPreferred = false;
+	
+	
+	/**
+	 * Sets whether local copies of articles are preferred over
+	 * articles from the server.
+	 *
+	 * @param isCachedArticlesPreferred true for local, false for server
+	 */
+	public void setCachedArticlesPreferred(boolean isCachedArticlesPreferred)
+	{
+		this.isCachedArticlesPreferred = isCachedArticlesPreferred;
+	}
+
+	/**
+	 * Checks if the client prefers articles from the local storage if available
+	 *
+	 * @return true, if local copies are preferred, false otherwise
+	 */
+	public boolean isCachedArticlesPreferred()
+	{
+		return isCachedArticlesPreferred;
+	}
+
+	/**
+	 * Instantiates a new distripedia client, and connects to the given URI.
+	 * 
+	 * @param url the url
+	 * @param localStorage the local storage
+	 */
+	public DistripediaClient(final URI url, final LocalStorage localStorage)
+	{
+		if (url == null || localStorage == null)
+		{
+			throw new IllegalArgumentException("Parameters cannot be null!");
+		}
+
+		this.localStorage = localStorage;
+
+		webSocket = new WebSocket(url);
+		webSocket.addOpenHandler(this);
+		webSocket.addCloseHandler(this);
+		webSocket.addMessageHandler(this);
+		webSocket.connect();
+	}
+
+	/**
+	 * Sends a request to the server to retrieve the article or tries to fetch
+	 * it from the localStorage if preferCachedArticles is set to true.
+	 * If the server can deliver the article an {@code ArticleEvent} will be
+	 * fired.
+	 * 
+	 * @param article the article
+	 */
+	public void getArticle(String article)
+	{
+		if (article == null)
+		{
+			throw new IllegalArgumentException("Article cannot be null!");
+		}
+
+		if (isCachedArticlesPreferred())
+		{
+			String content = localStorage.getItem(article);
+			
+			if (content != null)
+			{
+				// fire event specifying that article came from LS
+				handleArticleReceived(article, content, true);
+				return;
+			}
+		}
+
+		try
+		{
+			JSONObject request = new JSONObject();
+			request.put(Message.Key.ACTION, Message.Action.REQUEST_ARTICLE);
+			request.put(Message.Key.ARTICLE, article);
+
+			webSocket.send(request.toString());
+		} catch (JSONException e)
+		{
+			e.printStackTrace();
+		}	
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 109:119, 270:281.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.